readme.md
# @clearc2/headlamp

Generate express api documentation automagically.

## Requirements
This issues grep commands to search code. If you would like to use the source code searching feature, your system must
have [grep](https://www.gnu.org/software/grep/manual/grep.html) installed.

## Install
```
yarn add -D @clearc2/headlamp
```

Create a new `routes` directory.

```
server/
  routes/
  server.js
```

In your express api bootstrap file:

```js
// server/server.js

import express from 'express'
import bodyParser from 'body-parser'
import path from 'path'
import fs from 'fs'
import cors from 'cors'
import shine from '@clearc2/headlamp'

const app = express()
app.use(cors())
app.use(bodyParser.json())

// old routes here

shine(app, {
  title: 'IOP API Docs',
  description: String(fs.readFileSync(path.resolve(__dirname, '..', 'readme.md'))),
  routes: path.resolve(__dirname, 'routes'),
  src: path.resolve(__dirname, '..', 'src'),
  srcDir: '_headlamp',
  server: [
    path.resolve(__dirname),
    path.resolve(__dirname, '..', 'src', 'redux', 'server')
  ],
  hidePath: path.resolve(__dirname, '..'),
  headers: {
      Authorization: {
          value: 'foobar',
          help: 'jwt auth token'
      },
      'X-Custom-Thing': 'abc'
  },
  responses: [
    {
      title: 'Invalid request',
      status: 400,
      response: {
        errors: [
          {error: 'Invalid request'}
        ]
      }
    },
    {
      title: 'Server error',
      status: 500,
      response: {
        errors: [
          {error: 'Something went wrong.'}
        ]
      }
    }
  ]
})

const PORT = process.env.PORT || 5033
app.listen(PORT, () => {
  console.log('Dev Express server running at localhost:' + PORT)
})
```

After starting your mock server, you will have an api explorer at `http://localhost:<port>/_docs`.

## Configuration
The second argument to the `shine` function is a configuration object. All are optional.

#### `title`
Turns into the header of the api explorer.

#### `description`
General api description/documentation. Supports markdown.

#### `routes`
The directory to the route files. See route file documentation below.

#### `src`
The source code path. Can be a single path or array of paths.

#### `srcDir`
The name of directories _within_ your `src` where route files are located.

#### `server`
The mock server path. Can be a single path or array of paths.

#### `hidePath`
Removes the path from the file names in the api explorer.

#### `headers`
Headers to send for every request. See `params` documentation below for similar formatting.

#### `responses`
Default responses that will be associated with every file route. Useful for documenting error responses. Should be an array of
response objects. These will only be applied to routes defined via route files. See the `responses` documentation below.

## Route files
This package gives you a new way to define mock endpoints through route files. Route files can be nested in the `routes` directory. Example:

```
server/
  routes/
    auth-info.js
    companies/
      get-companies.js
      save-company.js
  server.js
```
The files will automatically get incorporated into the mock server as functional endpoints.

### Structure
Route files can either be a `.js` file that exports an object or a `.json` file. See this example route file.
```js
// server/routes/companies/employees.js

export default {
  path: '/companies/:companyUnid/employees',
  methods: ['GET'],
  title: 'Company employees',
  description: 'Fetches all active employees',
  response: {
    employees: [
      {
        name: 'John Doe'
      },
      {
        name: 'Bob Smith'
      }
    ]
  }
}
```

The equivalent json file of the above would look like below and behave identically.

```json
{
  "path": "/companies/:companyUnid/employees",
  "methods": ["GET"],
  "title": "Company employees",
  "description": "Fetches all active employees",
  "response": {
    "employees": [
      {
        "name": "John Doe"
      },
      {
        "name": "Bob Smith"
      }
    ]
  }
}
```

#### `path`
This must be a valid express route path.

#### `method`
A single http verb that this endpoint responds to. Ex. `POST`

#### `methods`
Routes can respond to multiple http methods. This must be an array of http verbs that this endpoint responds to.
Ex. `['GET', 'POST', 'PUT', 'PATCH', 'DELETE']`

#### `title`
Only used for documentation in the api explorer ui.

#### `description`
Only used for documentation in the api explorer ui. Supports markdown.

#### `params`
URL params are the tokens prefixed by `:` in the path. These params can be further defined:
```js
export default {
  path: '/companies/:companyUnid/employees',
  params: {
    companyUnid: 'abcdef'
  }
  // ...
}
```

The above will prepopulate this param's input in the api explorer. You can provide additional help text by using an object.
```js
export default {
  path: '/companies/:companyUnid/employees',
  params: {
    companyUnid: {
      value: 'abcdef',
      help: 'Company Unid'
    }
  }
  // ...
}
```
The `help` text will show up directly beneath the param's input in the api explorer.

#### `query`
Query params can be defined in the same way.
```js
export default {
  path: '/companies/:companyUnid/employees',
  query: {
    hiredAfter: {
      value: '01-01-2018',
      help: 'Format MM-DD-YYYY'
    }
  }
  // ...
}
```

#### `headers`
Headers to send with this request. Same format as `params` and `query`.

#### `payload`
`POST`, `PUT`, and `PATCH` methods allow for example payloads. These payloads will populate the payload input in the api explorer.
```js
export default {
  path: '/contact',
  methods: ['POST'],
  payload: {
    contact: {
      firstName: 'John',
      lastName: 'Doe'
    }
  }
  // ...
}
```

In the case of endpoints accepting both `POST` and `PUT` requests, you can define unique payloads for both.
```js
export default {
  path: '/contact',
  methods: ['POST', 'PUT'],
  payload: {
    POST: {
      contact: {
        firstName: 'John',
        lastName: 'Doe'
      }
    },
    PUT: {
      contact: {
        id: 'abcded',
        firstName: 'John',
        lastName: 'Doe'
      }
    }
  }
  // ...
}
```

#### `response`
Responses can be simple javascript objects...
```js
export default {
  path: '/people',
  methods: ['GET'],
  response: {
    people: {
      [
        {name: 'John Doe'},
        {name: 'Bob Smith'}
      ]
    }
  }
}
```

Or come from fixture files...
```js
export default {
  path: '/people',
  methods: ['GET'],
  response: require('./_people.json')
}
```

Or dynamically generated based on the request...

```js
export default {
  path: '/people',
  methods: ['GET', 'POST', 'PUT'],
  response: (req, res) => {
    if (req.method === 'POST') {
      return res.status(400).json({
        errors: [
          {detail: 'Something went wrong :('}
        ]
      })
    }

    if (req.method === 'PUT') {
      return res.status(500).json({
        errors: [
          {detail: 'CRITICAL FAILURE'}
        ]
      })
    }

    return res.status(200).json(require('./_people.json'))
  }
}
```

Unless a function is provided, a 200 status code will be sent.

#### `responses`
Instead of, or in conjunction with, the `response` key, you can provide static responses.

```js
export default {
  path: '/people',
  method: 'POST',
  responses: [
    {
      title: 'All good',
      description: 'Some **markdown** describing this response',
      status: 200,
      response: {
        foo: 'bar'
      }
    },
    () => ({
      title: 'Unauthorized',
      status: 401,
      response: require('./_bigData') // large fixture that doesn't need to be loaded into memory right away
    }),
    {
      title: 'No bueno',
      status: 500,
      response: {
        errors: [
          {error: 'Something went wrong :('}
        ]
      }
    }
  ]
}
```

These responses will be displayed in the api explorer. Because of that, they do not have access to the
express `req` or `res` objects. Zero arguments are passed to functional responses. If you need to add a delay or dynamically
generate a response based on the request, use `response`.

#### `globalHeaders`
Boolean to determine whether to merge the globally defined `headers` to this route. Defaults `true`.

#### `globalResponses`
Boolean to determine whether to concat the globally defined `responses` to this route. Defaults `true`.

### Response objects
Static responses must either be an object or a function that returns an object. Static response objects have the following fields:

##### `title`
The title of the static response.

##### `description`
The description of the static response. Supports markdown.

##### `status`
The status http status code.

##### `response`
The response payload.

## Response resolution
The following priority is used to determine which response to use.

- If activated response, use it
- If `response` object/function exists, use it
- If first response in `responses` exists, use it
- Sends error response

## Fixtures
Fixture files can be nested in the `routes` directory but their file names must be prefixed with an underscore
to not be interpreted as a route file.

```sh
server/
  routes/
    people/
      get-people/
        _people.json     # will not be interpreted as a route file
        route.js         # will be interpreted as a route file
      save-person.js     # will be interpreted as a route file
      update-person.json # will be interpreted as a route file
```

## Explorer features

### Search
Searching in the frontend is case-insensitive. You can omit path variables and simply provide a colon. Ex.

```
/companies/:/employees/:/address
```

The above would match:


```
/companies/:companyUnid/employees/:unid/address
```


CHANGELOG.md
## [2.5.1](https://github.com/ClearC2/headlamp/compare/v2.5.0...v2.5.1) (2023-10-25)


### Bug Fixes

* **custom-response:** fix possible null ([fc24288](https://github.com/ClearC2/headlamp/commit/fc24288914114c60ad562e082cf50dea7c0a0eea))

# [2.5.0](https://github.com/ClearC2/headlamp/compare/v2.4.0...v2.5.0) (2023-09-21)


### Features

* add copy response ([b7e90b8](https://github.com/ClearC2/headlamp/commit/b7e90b838b29bca3d50dfbc220cf6c220adfb7c2))

# [2.4.0](https://github.com/ClearC2/headlamp/compare/v2.3.2...v2.4.0) (2023-03-08)


### Features

* support cjs files ([08018b3](https://github.com/ClearC2/headlamp/commit/08018b30d9b220b7bf3644ad3d15f2229c23df25))

## [2.3.2](https://github.com/ClearC2/headlamp/compare/v2.3.1...v2.3.2) (2023-02-24)


### Bug Fixes

* make it faster ([417cedf](https://github.com/ClearC2/headlamp/commit/417cedf96adf9ac557e82aec0356c556dba49bd7))

## [2.3.1](https://github.com/ClearC2/headlamp/compare/v2.3.0...v2.3.1) (2023-02-24)


### Bug Fixes

* routes reference ([#59](https://github.com/ClearC2/headlamp/issues/59)) ([fe76ee1](https://github.com/ClearC2/headlamp/commit/fe76ee18cf93b68ec396e0d2f81c78c6b0b90848))

# [2.3.0](https://github.com/ClearC2/headlamp/compare/v2.2.1...v2.3.0) (2023-02-24)


### Features

* add srcDir support ([aa9690d](https://github.com/ClearC2/headlamp/commit/aa9690d38e892ab229c63a8d2804979057e5b213))

## [2.2.1](https://github.com/ClearC2/headlamp/compare/v2.2.0...v2.2.1) (2023-02-09)


### Bug Fixes

* default custom route title ([2f93ef7](https://github.com/ClearC2/headlamp/commit/2f93ef7e459a2685b56c2ecd30171f2479b2ca62))

# [2.2.0](https://github.com/ClearC2/headlamp/compare/v2.1.1...v2.2.0) (2022-07-14)


### Features

* add har support ([f1bee8c](https://github.com/ClearC2/headlamp/commit/f1bee8c0557bef15ba36a2124a68b7c67b32be0b))

## [2.1.1](https://github.com/ClearC2/headlamp/compare/v2.1.0...v2.1.1) (2020-11-24)


### Bug Fixes

* **api:** fix host in route link header ([0cf8940](https://github.com/ClearC2/headlamp/commit/0cf89404071444944c87107e024bf0f6cc792f69))

# [2.1.0](https://github.com/ClearC2/headlamp/compare/v2.0.3...v2.1.0) (2020-11-24)


### Features

* **api:** add route link to headers in api responses ([804ef56](https://github.com/ClearC2/headlamp/commit/804ef564be9e292262a82ce273ca282825f846dd))

## [2.0.3](https://github.com/ClearC2/headlamp/compare/v2.0.2...v2.0.3) (2020-02-12)


### Bug Fixes

* **deps:** make react packages not required ([#16](https://github.com/ClearC2/headlamp/issues/16)) ([17ceb69](https://github.com/ClearC2/headlamp/commit/17ceb69d78235ac4a771df590fc1576853dce9e1))

## [2.0.2](https://github.com/ClearC2/headlamp/compare/v2.0.1...v2.0.2) (2019-07-05)


### Bug Fixes

* fix bug when attempting to get responses from non headlamp route ([509b6c9](https://github.com/ClearC2/headlamp/commit/509b6c9))

## [2.0.1](https://github.com/ClearC2/headlamp/compare/v2.0.0...v2.0.1) (2019-06-25)


### Bug Fixes

* move [@hot-loader](https://github.com/hot-loader) to dev deps ([8f1a660](https://github.com/ClearC2/headlamp/commit/8f1a660))

# [2.0.0](https://github.com/ClearC2/headlamp/compare/v1.3.0...v2.0.0) (2019-06-25)


### Performance Improvements

* upgrade deps ([55ecc8a](https://github.com/ClearC2/headlamp/commit/55ecc8a))


### BREAKING CHANGES

* move to @clearc2 scope

# [1.3.0](https://github.com/ClearC2/headlamp/compare/v1.2.0...v1.3.0) (2019-04-02)


### Features

* add custom temporary responses ([#9](https://github.com/ClearC2/headlamp/issues/9)) ([d7b404e](https://github.com/ClearC2/headlamp/commit/d7b404e))

# [1.2.0](https://github.com/ClearC2/headlamp/compare/v1.1.2...v1.2.0) (2019-03-29)


### Features

* add ability to filter on http method ([#8](https://github.com/ClearC2/headlamp/issues/8)) ([75b6f33](https://github.com/ClearC2/headlamp/commit/75b6f33))

## [1.1.2](https://github.com/ClearC2/headlamp/compare/v1.1.1...v1.1.2) (2019-03-29)


### Bug Fixes

* **api:** remove last modified ([#7](https://github.com/ClearC2/headlamp/issues/7)) ([4c11a6b](https://github.com/ClearC2/headlamp/commit/4c11a6b))

## [1.1.1](https://github.com/ClearC2/headlamp/compare/v1.1.0...v1.1.1) (2019-03-20)


### Bug Fixes

* upgrade deps for security reasons ([bc1a997](https://github.com/ClearC2/headlamp/commit/bc1a997))

# [1.1.0](https://github.com/ClearC2/headlamp/compare/v1.0.0...v1.1.0) (2019-03-13)


### Features

* **headers:** add headers support ([#5](https://github.com/ClearC2/headlamp/issues/5)) ([60ebde2](https://github.com/ClearC2/headlamp/commit/60ebde2))

# 1.0.0 (2018-07-31)


### Features

* **api:** added server filename as res header ([9ca89ff](https://github.com/ClearC2/headlamp/commit/9ca89ff))


test-api/server.js
import express from 'express'
import bodyParser from 'body-parser'
import path from 'path'
import cors from 'cors'
import fs from 'fs'
import document from '../src/api-route-provider'

const app = express()
app.use(cors())
app.use(bodyParser.json())

app.put('/ticket/2.0/workorder/update', function (req, res) {
  return res.status(201).json({
    'message': 'Work order updated successfully.'
  })
})

app.put('/companies/:unid/workorder/:id/foobar', function (req, res) {
  return res.status(201).json({
    'message': 'Work order updated successfully.'
  })
})

document(app, {
  routes: path.resolve(__dirname, 'routes'),
  src: path.resolve(__dirname, '..', 'test-src'),
  srcDir: '_headlamp',
  server: path.resolve(__dirname, '..', 'test-api'),
  // title: 'Example API Docs',
  description: String(fs.readFileSync(path.resolve(__dirname, '..', 'readme.md'))),
  hidePath: path.resolve(__dirname, '..'),
  headers: {
    authorization: 'bar'
  },
  responses: [
    {
      title: 'Error',
      description: 'Typical error',
      status: 400,
      response: {
        errors: [
          {detai: 'Something went wrong :('}
        ]
      }
    },
    () => ({
      title: 'Critical Error',
      description: 'Typical error',
      status: 500,
      response: {
        errors: [
          {detai: 'Something went realllly wrong >:('}
        ]
      }
    })
  ]
})

const PORT = process.env.PORT || 5033
app.listen(PORT, () => {
  console.log('Dev Express server running at localhost:' + PORT) // eslint-disable-line
})
